warehouse/redshift/redshift.go

```diff
@@ -296,7 +307,17 @@ func (rs *HandleT) loadTable(tableName string, columnMap map[string]string, buck
 		primaryKey = column
 	}
 
-	sqlStatement = fmt.Sprintf(`DELETE FROM %[1]s."%[2]s" using %[1]s."%[3]s" _source where _source.%[4]s = %[1]s.%[2]s.%[4]s`, rs.Namespace, tableName, stagingTableName, primaryKey)
+	partitionKey := "id"
+	if column, ok := partitionKeyMap[tableName]; ok {
+		partitionKey = column
+	}
+
+	var additionalJoinClause string
+	if tableName == warehouseutils.DiscardsTable {
+		additionalJoinClause = fmt.Sprintf(`AND _source.%[3]s = %[1]s.%[2]s.%[3]s`, rs.Namespace, tableName, "column_name")
+	}
+
+	sqlStatement = fmt.Sprintf(`DELETE FROM %[1]s."%[2]s" using %[1]s."%[3]s" _source where (_source.%[4]s = %[1]s.%[2]s.%[4]s %[5]s)`, rs.Namespace, tableName, stagingTableName, primaryKey, additionalJoinClause)
```

Comment:
Delete should not include additionalJoinClause, and delete everything for a row_id instead. In a scenario where you already have discarded values for an id X and columns A, B, C, and later you need to replace with discarded values for the same id and columns B, C, D, you would end up with all for columns A, B, C, D, with A corresponding to the previous version of X, which will have been replaced in the main table.

The join should also include the table_name to avoid the rare cases where the same id value appears in different tables.

The same applies to BigQuery & snowflake

Reply:
@fxenik removed `column_name` from `additionalJoinClause` and added `table_name` to it delete only rows corresponding to that table in rare case where row_id is same across table as you pointed out above.